runtime/runtime-params-estimator/src/function_call.rs

```diff
@@ -0,0 +1,150 @@
+use near_primitives::config::VMConfig;
+use near_primitives::contract::ContractCode;
+use near_primitives::profile::ProfileData;
+use near_primitives::runtime::fees::RuntimeFeesConfig;
+use near_primitives::types::{CompiledContractCache, ProtocolVersion};
+use near_store::{create_store, StoreCompiledContractCache};
+use near_vm_logic::mocks::mock_external::MockedExternal;
+use near_vm_runner::{run_vm, VMKind};
+use nearcore::get_store_path;
+use std::fmt::Write;
+use std::sync::Arc;
+
+#[allow(dead_code)]
+fn test_function_call(metric: GasMetric, vm_kind: VMKind) {
+    let mut xs = vec![];
+    let mut ys = vec![];
+    const REPEATS: i32 = 50;
+    for method_count in vec![5, 20, 30, 50, 100, 200, 1000] {
+        let contract = make_many_methods_contact(method_count);
+        let cost =
+            compute_function_call_cost(metric, vm_kind, REPEATS, &contract);
+        println!("{:?} {:?} {} {}", vm_kind, metric, method_count, cost / (REPEATS as u64));
+        xs.push(contract.get_code().len() as u64);
+        ys.push(cost / (REPEATS as u64));
+    }
+
+    // Regression analysis only makes sense for additive metrics.
+    if metric == GasMetric::Time {
+        return;
+    }
+
+    let (cost_base, cost_byte, _) = least_squares_method(&xs, &ys);
+
+    println!(
+        "{:?} {:?} function call base {} gas, per byte {} gas",
+        vm_kind, metric,
+        ratio_to_gas_signed(metric, cost_base),
+        ratio_to_gas_signed(metric, cost_byte),
+    );
+}
+
+#[test]
+fn test_function_call_time() {
+    // Run with
+    // cargo test --release --lib function_call::test_function_call_time -- --exact --nocapture
+    test_function_call(GasMetric::Time, VMKind::Wasmer0);
+    test_function_call(GasMetric::Time, VMKind::Wasmer1);
+    test_function_call(GasMetric::Time, VMKind::Wasmtime);
+
+}
+
+#[test]
+fn test_function_call_icount() {
+    // Use smth like
+    // CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_RUNNER=./runner.sh \
+    // cargo test --release --features no_cpu_compatibility_checks \
+    // --lib function_call::test_function_call_icount -- --exact --nocapture
+    // Where runner.sh is
+    // /host/nearcore/runtime/runtime-params-estimator/emu-cost/counter_plugin/qemu-x86_64 \
+    // -cpu Westmere-v1 -plugin file=/host/nearcore/runtime/runtime-params-estimator/emu-cost/counter_plugin/libcounter.so $@
+    test_function_call(GasMetric::ICount, VMKind::Wasmer0);
+    test_function_call(GasMetric::ICount, VMKind::Wasmer1);
+    test_function_call(GasMetric::ICount, VMKind::Wasmtime);
+}
+
+fn make_many_methods_contact(method_count: i32) -> ContractCode {
+    let mut methods = String::new();
+    for i in 0..method_count {
+        write!(
+            &mut methods,
+            "
+            (export \"hello{}\" (func {}))
+              (func (;{};)
+                i32.const {}
+                drop
+                return
+              )
+            ", i, i, i, i)
```

Comment:
```suggestion
            r#"
            (export "hello{i}" (func {i}))
              (func (;{i};)
                i32.const {i}
                drop
                return
              )
            "#, i=i)
```

although I guess it's better to use some wat sugar here to not repeat the i in the first place:

```rust
        write!(
            &mut methods,
            r#"
            (func (export "hello{i}")
                i32.const {i}
                drop
                return
            )
            "#,
            i = i
        )

```